File: twoPointers/jsSolutions/02-threeSum-medium.js
/*
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]]
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.
*/

const threeSum = (nums) => {
  nums.sort((a, b) => a - b)
  let res = {}
  let len = nums.length - 1

  for (let i = 0; i <= len - 2; i++) {
    let left = i + 1
    let right = len

    while (left < right) {
      let set = [nums[i], nums[left], nums[right]]
      let sum = set.reduce((acc, n) => acc + n, 0)
      if (sum === 0) {
        let str = `${set}`
        if (!(str in res)) res[str] = set
      }
      if (sum <= 0) left++
      else right--
    }
  }

  return Object.values(res)
}


// nums = [-1, 0, 1, 2, -1, -4]
// console.log(threeSum(nums))
// nums2 = [0, 0, 0, 0]
// console.log(threeSum(nums2))
// nums3 = [-2, 0, 1, 1, 2]
// console.log(threeSum(nums3))
nums4 = [-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4]
console.log(threeSum(nums4))

File: twoPointers/pySolutions/02-three_sum-medium.py
'''
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]]
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.
'''


def three_sum(nums):
    nums.sort()
    res = {}
    end = len(nums) - 1

    for i in range(0, end):
        l = i+1
        r = end

        while l < r:
            numset = [nums[i], nums[l], nums[r]]
            if sum(numset) == 0:
                res[str(numset)] = numset
                pass
            if sum(numset) < 0:
                l += 1
            else:
                r -= 1

    return list(res.values())


nums = [-1, 0, 1, 2, -1, -4]
print(three_sum(nums))
nums2 = [0, 0, 0, 0]
print(three_sum(nums2))
nums3 = [-2, 0, 1, 1, 2]
print(three_sum(nums3))
nums4 = [-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4]
print(three_sum(nums4))
